frontend/src/metabase/services.unit.spec.ts

```diff
@@ -0,0 +1,199 @@
+import fetchMock from "fetch-mock";
+import { createEntitiesState } from "__support__/store";
+
+import { defer } from "metabase/lib/promise";
+import { getMetadata } from "metabase/selectors/metadata";
+
+import type {
+  Card,
+  DashboardId,
+  DashCardId,
+  UnsavedCard,
+} from "metabase-types/api";
+import {
+  createMockCard,
+  createMockUnsavedCard,
+  createMockDataset,
+  createMockStructuredDatasetQuery,
+} from "metabase-types/api/mocks";
+import {
+  createSampleDatabase,
+  ORDERS_ID,
+  SAMPLE_DB_ID,
+} from "metabase-types/api/mocks/presets";
+import { createMockState } from "metabase-types/store/mocks";
+
+import Question from "metabase-lib/Question";
+
+import { runQuestionQuery } from "./services";
+
+const MOCK_QUERY = createMockStructuredDatasetQuery({
+  database: SAMPLE_DB_ID,
+  query: {
+    "source-table": ORDERS_ID,
+  },
+});
+
+function createMockMetadata(card?: Card) {
+  const database = createSampleDatabase();
+  const state = createMockState({
+    entities: createEntitiesState({
+      databases: [database],
+      questions: card ? [card] : [],
+    }),
+  });
+  return getMetadata(state);
+}
+
+type DashboardAwareCard = Card & {
+  dashboardId?: DashboardId;
+  dashcardId?: DashCardId;
+};
+
+function createMockSavedQuestion(card?: Partial<DashboardAwareCard>) {
+  const savedCard = createMockCard({ dataset_query: MOCK_QUERY, ...card });
+  return createMockMetadata(savedCard).question(savedCard.id) as Question;
+}
```

Comment:
So in fact, a saved card can carry dashboard and dashcard IDs. I'm not sure if it's worth adding these properties to the `Card` type — this seems like a workaround for dashboard <—> query builder interactions and should probably be solved on a higher level 🤔